src/librustc_typeck/collect.rs

```diff
@@ -1699,9 +1812,26 @@ fn find_opaque_ty_constraints(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
     }
 }
 
+crate fn is_infer_ty(ty: &hir::Ty) -> bool {
+    match &ty.kind {
+        hir::TyKind::Infer => true,
+        hir::TyKind::Slice(ty) | hir::TyKind::Array(ty, _) => is_infer_ty(ty),
+        hir::TyKind::Tup(tys)
+            if !tys.is_empty()
+                && tys.iter().all(|ty| match ty.kind {
+                    hir::TyKind::Infer => true,
+                    _ => false,
+                }) =>
+        {
+            true
+        }
+        _ => false,
```

Comment:
can't infer types also be in fields of generic structs?

Reply:
They can, but this is only run in a hacky way to provide suggestions when writing `fn foo() -> _` using inference from the body, I didn't want to dig too deeply in that direction before getting the general case to work correctly.